zfs: build pyzfs Python module

lschuermann opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
This changes the ZFS on Linux derivation to also build the `pyzfs` Python module, which exports the `libzfs_core` bindings. Those bindings are useful to get more direct access to ZFS primitives. For instance, I needed to use this library to resolve some issues around ZFS encryption: https://github.com/openzfs/zfs/issues/12614#issuecomment-1936932729

### Open Questions

I'm not sure whether this is the "correct" approach for packing `pyzfs`. I would prefer to use `buildPythonPackage`, as this currently does not set `PYTHONPATH` correctly, doesn't do import checks, needs to have its `site-packages` install prefix supplied manually, etc.

However, the OpenZFS repository uses autotools to template parts of the Python source, such as the `setup.py` script: https://github.com/openzfs/zfs/blob/8f2f6cd2ac688916adb2caf979daf95365ccb48f/contrib/pyzfs/setup.py.in

I see a couple of potential solutions here, each one of which I'm not sure how well it would work or integrate into the larger Nixpkgs ecosystem:
1. Leave this as is. Its sufficient to be able to use `pyzfs`, provided the `PYTHONPATH` and `LD_LIBRARY_PATH` are amended manually.
2. Try to hack the main ZFS package into behaving like a Python package derivation. I do not know how to do this, and its likely problematic from a maintenance perspective.
3. Only run the `unpack` and `configure` phases of the ZFS derivation, and then use this intermediate result as the source for a new `buildPythonPackage` derivation.
   - Is this easily possible and maintainable?
   - Where do we put this new package? Should it be an attribute on the main ZFS package? Or located under `python3Packages`? For the latter, how do we clearly communicate that this is the `pyzfs` library built specifically for / from the same source as the main ZFS derivation?

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

lschuermann committed:
pkgs/os-specific/linux/zfs: build pyzfs Python module

lschuermann commented on pkgs/os-specific/linux/zfs/generic.nix:
```diff
@@ -134,6 +136,7 @@ let
       "--localstatedir=/var"
       "--enable-systemd"
       "--enable-pam"
+      (if enablePython then "--enable-pyzfs" else "")
```
```suggestion
      (lib.optionalString enablePython "--enable-pyzfs")
```

lschuermann commented on pkgs/os-specific/linux/zfs/generic.nix:
```diff
@@ -111,7 +111,9 @@ let
-      ++ optional (buildUser && enablePython) python3;
+      ++ optionals (buildUser && enablePython) (with python3Packages; [
+        python3 packaging setuptools cffi
+      ]);
```
Thanks, changed!

lschuermann commented:
@amarshall Thank you for the review, and sorry for getting back to this after such long delay!

I went with the recommended `toPythonModule` route to expose this package as part of the `pythonPackages` attribute set. Now you should be able to use it through the `python3Packages.libzfs_core` attribute without any additional environment variables being set. I needed to add a patch to change the `cffi` binding's `dlopen` call to refer to the `zfs-user` derivation directly, as in contrast to other packages using `cffi`, this one won't have the required shared libraries copied into the Python derivation itself.

amarshall commented on pkgs/os-specific/linux/zfs/generic.nix:
```suggestion
        --replace-fail 'self._ffi.dlopen(self._libname' \
```

amarshall commented on pkgs/os-specific/linux/zfs/generic.nix:
Could you add the pythonimportcheck hook? (I tried to add comment to the file, but GitHub errored)

amarshall reviewed (commented):
Needs rebase to resolve merge conflicts as well.

lschuermann commented:
@amarshall Thanks for the review! I think this PR is still valuable to get merged eventually. I rebased and implemented your suggestions; my basic local tests suggest that this still works!

adamcstephens commented:
I agree with your initial assessment that this isn't really an idiomatic python package. The output is effecitvely the same as zfs userspace, with a python site-packages included.

Does this need to match the exact version of zfs that is running on the system? Some possible paths:

- Don't use the zfs package at all. Re-declare the src in a python package just for pyzfs. If version mismatches are ok, this is maybe the best option.
- Embed a standalone python package inside the zfs package, maybe in passthru.pyzfs or something. It can then be a `buildPythonPackage` but can be passed the src, etc.

lschuermann commented:
@adamcstephens I don't disagree with that, but I don't think I have the time or energy to invest in any significant transformations of the package.

> Does this need to match the exact version of zfs that is running on the system?

I believe that is true, yes.

> - Don't use the zfs package at all. Re-declare the src in a python package just for pyzfs. If version mismatches are ok, this is maybe the best option.
> - Embed a standalone python package inside the zfs package, maybe in passthru.pyzfs or something. It can then be a buildPythonPackage but can be passed the src, etc.

I don't think that either of those will be easy to accomplish in a clean way, because `pyzfs` is effectively a build artifact of the regular ZFS autotools-based build toolchain.

The only possible way that I think we can somewhat easily make this a more idiomatic Python package is by taking the output from the ZFS derivation and removing all non-Python elements from it. That's effectively what we'd have to do for a dedicated `pyzfs` derivation anyways, and it would not be sustainable to duplicate all of the ZFS build steps between those two packages.

But, skimming over other `toPythonModule` packages, I think this PR is not introducing a new paradigm. Other packages seems to be in a similar situation, and effectively do the same thing that this package is confusing.

Can you give me some concrete guidance on how to proceed? As I'm fairly stretched on time, I don't want to sink hours into something that will ultimately be unmergeable.

adamcstephens commented:
If pyzfs needs to match the system zfs version, then you'll need to consider that we support multiple zfs versions. This means we'll need a matrix of python and zfs versions, which this currently doesn't support. I'm not opposed to extending the zfs package as-is, even if python won't need the other bits.

adamcstephens commented on pkgs/os-specific/linux/zfs/generic.nix:
```diff
@@ -17,6 +17,7 @@ let
       zlib,
       libuuid,
       python3,
+      python3Packages,
```
If we're adding python3Packages, we should use it for python itself. e.g. drop the `python3` argument above this one, and replace references to it to `python3Packages.python`.

lschuermann replied:
Changed, thanks!

adamcstephens commented on pkgs/top-level/python-packages.nix:
```diff
@@ -8949,6 +8949,12 @@ self: super: with self; {
+    pkgs.zfs.override {
+      enablePython = true;
+    }
+  );
```
```suggestion
  libzfs_core_2_3 = toPythonModule (
    pkgs.zfs_2_3.override {
      enablePython = true;
      python3Packages = self;
    }
  );
```

I'm not sure about this package name, but something to enable the matrix seems important. Upstream doesn't appear to actually publish this package in pypi...

lschuermann replied:
`libzfs_core_<version>` seems good to me. An alternative could be to introduce a new attrset per ZFS version like `zfs_<version>.libzfs_core`, but I don't know whether that may not break anything.

lschuermann replied:
Changed it to be two separate packages for the currently supported set of ZFS versions.
